toolchain/check/deduce.cpp

```diff
@@ -0,0 +1,196 @@
+      Add(param, arg, needs_substitution);
+    }
+  }
+
+  auto AddBlock(SemIR::InstBlockId params, llvm::ArrayRef<SemIR::InstId> args,
+                bool needs_substitution) -> void {
+    AddBlock(context.inst_blocks().Get(params), args, needs_substitution);
+  }
+
+  auto AddBlock(SemIR::InstBlockId params, SemIR::InstBlockId args,
+                bool needs_substitution) -> void {
+    AddBlock(context.inst_blocks().Get(params), context.inst_blocks().Get(args),
+             needs_substitution);
+  }
+
+  struct PendingDeduction {
+    SemIR::InstId param;
+    SemIR::InstId arg;
+    bool needs_substitution;
+  };
+  Context& context;
+  llvm::SmallVector<PendingDeduction> deductions;
+};
+}  // namespace
+
+static auto NoteGenericHere(Context& context, SemIR::GenericId generic_id,
+                            Context::DiagnosticBuilder& diag) -> void {
+  CARBON_DIAGNOSTIC(DeductionGenericHere, Note,
+                    "While deducing parameters of generic declared here.");
+  diag.Note(context.generics().Get(generic_id).decl_id, DeductionGenericHere);
+}
+
+auto DeduceGenericCallArguments(Context& context, Parse::NodeId node_id,
+                                SemIR::GenericId generic_id,
+                                SemIR::SpecificId enclosing_specific_id,
+                                SemIR::InstBlockId implicit_params_id,
+                                SemIR::InstBlockId params_id,
+                                SemIR::InstId self_id,
+                                llvm::ArrayRef<SemIR::InstId> arg_ids)
+    -> SemIR::SpecificId {
+  DeductionWorklist worklist = {.context = context};
+
+  // TODO: Perform deduction for type of self
+  static_cast<void>(implicit_params_id);
+  static_cast<void>(self_id);
+
+  // Copy any outer generic arguments from the specified instance and prepare to
+  // substitute them into the function declaration.
+  llvm::SmallVector<SemIR::InstId> results;
+  llvm::SmallVector<Substitution> substitutions;
+  if (enclosing_specific_id.is_valid()) {
+    auto args = context.inst_blocks().Get(
+        context.specifics().Get(enclosing_specific_id).args_id);
+    results.assign(args.begin(), args.end());
+
+    // TODO: Subst is linear in the length of the substitutions list. Change it
+    // so we can pass in an array mapping indexes to substitutions instead.
+    substitutions.reserve(args.size());
+    for (auto [i, subst_inst_id] : llvm::enumerate(args)) {
+      substitutions.push_back(
+          {.bind_id = SemIR::CompileTimeBindIndex(i),
+           .replacement_id = context.constant_values().Get(subst_inst_id)});
+    }
+  }
+  auto first_deduced_index = SemIR::CompileTimeBindIndex(results.size());
+
+  worklist.AddBlock(params_id, arg_ids, /*needs_substitution=*/true);
+
+  results.resize(context.inst_blocks()
+                     .Get(context.generics().Get(generic_id).bindings_id)
+                     .size(),
+                 SemIR::InstId::Invalid);
+
```

Comment:
```suggestion
  results.resize(context.inst_blocks()
                     .Get(context.generics().Get(generic_id).bindings_id)
                     .size(),
                 SemIR::InstId::Invalid);

  worklist.AddBlock(params_id, arg_ids, /*needs_substitution=*/true);

```

(This is very optional, I just find it slightly clearer to group together the code that's setting up `results` for the main loop)

Reply:
Thanks, I agree that's clearer. Done.